The job **`dist-aarch64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13864447261/job/38800397834) [(plain)](https://github.com/rust-lang-ci/rust/commit/06727622a9092b30da74662a5f922a89be95206b/checks/38800397834/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] alloc test:false 5.657
[RUSTC-TIMING] hashbrown test:false 1.196
error: linking with `link.exe` failed: exit code: 1120
  |
  = note: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.43.34808\\bin\\HostX64\\arm64\\link.exe" "/DEF:C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\rustcRtM1o2\\lib.def" "/NOLOGO" "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\rustcRtM1o2\\symbols.o" "<1 object files omitted>" "<sysroot>-std\\arm64ec-pc-windows-msvc\\release\\deps\\std-23f1be5f370902f3.8fvzpx0f4fcm9u43hto3ssbg7.rcgu.rmeta" "<1 object files omitted>" "kernel32.lib" "kernel32.lib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage2-std\\arm64ec-pc-windows-msvc\\release\\deps/{libpanic_unwind-64589e62d02ec034.rlib,libwindows_targets-8318bc46163d768b.rlib,librustc_demangle-f9c710718bc4d403.rlib,libstd_detect-792ca7a0e60cb8a5.rlib,libhashbrown-be97f4187d164e69.rlib,librustc_std_workspace_alloc-e1e1ee2fa73b7012.rlib,libunwind-d57960c744869a6e.rlib,libcfg_if-4d55007f2c3a9957.rlib,liballoc-f8c9963fe2fe5e7a.rlib,librustc_std_workspace_core-48e39f4c66293531.rlib,libcore-234432e35e59ce43.rlib,libcompiler_builtins-7551276725dade63.rlib}.rlib" "ntdll.lib" "userenv.lib" "ws2_32.lib" "dbghelp.lib" "/defaultlib:libcmt" "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\rustcRtM1o2\\bcryptprimitives.dll_imports.lib" "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\rustcRtM1o2\\api-ms-win-core-synch-l1-2-0.dll_imports.lib" "/machine:arm64ec" "softintrin.lib" "/NXCOMPAT" "/LIBPATH:C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.43.34808\\atlmfc\\lib\\arm64" "/LIBPATH:<sysroot>-std\\arm64ec-pc-windows-msvc\\release\\build\\compiler_builtins-475aa472ac72abe7\\out" "/OUT:<sysroot>-std\\arm64ec-pc-windows-msvc\\release\\deps\\std-23f1be5f370902f3.dll" "/OPT:REF,ICF" "/DLL" "/IMPLIB:<sysroot>-std\\arm64ec-pc-windows-msvc\\release\\deps\\std-23f1be5f370902f3.dll.lib" "/DEBUG" "/PDBALTPATH:%_PDB%"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: LINK : warning LNK4279: function '__imp_SetFilePointerEx' mapped to different icall push thunks: '$iexit_thunk$cdecl$i8$i8m8i8i8' (in 'libucrt.lib(_flsbuf.obj)') and '$iexit_thunk$cdecl$i8$i8i8i8i8' (in 'std-23f1be5f370902f3.std.b58f7a683fc58c1b-cgu.0.rcgu.o')␍
          LINK : warning LNK4279: function '__imp_SetFilePointerEx' mapped to different icall push thunks: '$iexit_thunk$cdecl$i8$i8m8i8i8' (in 'libucrt.lib(lseek.obj)') and '$iexit_thunk$cdecl$i8$i8i8i8i8' (in 'std-23f1be5f370902f3.std.b58f7a683fc58c1b-cgu.0.rcgu.o')␍
          LINK : warning LNK4279: function '__imp_SetFilePointerEx' mapped to different icall push thunks: '$iexit_thunk$cdecl$i8$i8m8i8i8' (in 'libucrt.lib(lseek.obj)') and '$iexit_thunk$cdecl$i8$i8i8i8i8' (in 'std-23f1be5f370902f3.std.b58f7a683fc58c1b-cgu.0.rcgu.o')␍
          LINK : warning LNK4279: function '__imp_SetFilePointerEx' mapped to different icall push thunks: '$iexit_thunk$cdecl$i8$i8m8i8i8' (in 'libucrt.lib(lseek.obj)') and '$iexit_thunk$cdecl$i8$i8i8i8i8' (in 'std-23f1be5f370902f3.std.b58f7a683fc58c1b-cgu.0.rcgu.o')␍
          LINK : warning LNK4279: function '__imp_SetFilePointerEx' mapped to different icall push thunks: '$iexit_thunk$cdecl$i8$i8m8i8i8' (in 'libucrt.lib(lseek.obj)') and '$iexit_thunk$cdecl$i8$i8i8i8i8' (in 'std-23f1be5f370902f3.std.b58f7a683fc58c1b-cgu.0.rcgu.o')␍
          LINK : warning LNK4279: function '__imp_SetFilePointerEx' mapped to different icall push thunks: '$iexit_thunk$cdecl$i8$i8m8i8i8' (in 'libucrt.lib(lseek.obj)') and '$iexit_thunk$cdecl$i8$i8i8i8i8' (in 'std-23f1be5f370902f3.std.b58f7a683fc58c1b-cgu.0.rcgu.o')␍
          LINK : warning LNK4279: function '__imp_SetFilePointerEx' mapped to different icall push thunks: '$iexit_thunk$cdecl$i8$i8m8i8i8' (in 'libucrt.lib(lseek.obj)') and '$iexit_thunk$cdecl$i8$i8i8i8i8' (in 'std-23f1be5f370902f3.std.b58f7a683fc58c1b-cgu.0.rcgu.o')␍
          LINK : warning LNK4279: function '__imp_SetFilePointerEx' mapped to different icall push thunks: '$iexit_thunk$cdecl$i8$i8m8i8i8' (in 'libucrt.lib(lseek.obj)') and '$iexit_thunk$cdecl$i8$i8i8i8i8' (in 'std-23f1be5f370902f3.std.b58f7a683fc58c1b-cgu.0.rcgu.o')␍
          LINK : warning LNK4279: function '__imp_SetFilePointerEx' mapped to different icall push thunks: '$iexit_thunk$cdecl$i8$i8m8i8i8' (in 'libucrt.lib(lseek.obj)') and '$iexit_thunk$cdecl$i8$i8i8i8i8' (in 'std-23f1be5f370902f3.std.b58f7a683fc58c1b-cgu.0.rcgu.o')␍
             Creating library D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2-std\arm64ec-pc-windows-msvc\release\deps\std-23f1be5f370902f3.dll.lib and object D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2-std\arm64ec-pc-windows-msvc\release\deps\std-23f1be5f370902f3.dll.exp␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN100_$LT$compiler_builtins..math..libm..support..big..u256$u20$as$u20$core..ops..bit..Shl$LT$u32$GT$$GT$3shl17h95c2898d70ca4b5cE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins3int3big71_$LT$impl$u20$compiler_builtins..int..traits..HInt$u20$for$u20$u128$GT$14zero_widen_mul17h1cc840538886c889E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins3int3big71_$LT$impl$u20$compiler_builtins..int..traits..HInt$u20$for$u20$u128$GT$8widen_hi17hdd55af3779e2e2aeE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins3int3big71_$LT$impl$u20$compiler_builtins..int..traits..HInt$u20$for$u20$u128$GT$9widen_mul17h22d56e7ea3c7f007E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins3int3mul9__muloti417h4efef3eac629a2b5E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins3int6addsub15__rust_u128_add17hdf2461693d110915E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins3int6addsub15__rust_u128_sub17hde5637be46d7c84aE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins3mem6memcmp17h521a95e7ea70a4c9E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins4math4libm7support3big91_$LT$impl$u20$compiler_builtins..math..libm..support..int_traits..HInt$u20$for$u20$u128$GT$8widen_hi17hb6abfafbc4327646E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins4math4libm7support3big91_$LT$impl$u20$compiler_builtins..math..libm..support..int_traits..HInt$u20$for$u20$u128$GT$9widen_mul17haedfe9d288e19bbfE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins5float3cmp7__gtdf217h046c57f13929951dE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins5float3cmp7__gtsf217hb9089fa0af6fa5b2E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins5float3cmp7__ledf217h3f7f86fe06f7abb7E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins5float3cmp7__lesf217h9c5e8b104ceb4757E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins5float3cmp7__ltdf217h7d3e8e28db2979b4E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins5float3cmp7__ltsf217hfad6d25f36c0fd06E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins5float3cmp7__nedf217hb2370d6abad9f334E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN17compiler_builtins5float3cmp7__nesf217h5eace41323bbc5dbE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN58_$LT$u8$u20$as$u20$compiler_builtins..int..traits..Int$GT$11rotate_left17h8a6cfefa9a59a3d5E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN58_$LT$u8$u20$as$u20$compiler_builtins..int..traits..Int$GT$13leading_zeros17h13c2f57921af9382E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN59_$LT$u16$u20$as$u20$compiler_builtins..int..traits..Int$GT$11rotate_left17h080fdbd66f6f1134E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN59_$LT$u16$u20$as$u20$compiler_builtins..int..traits..Int$GT$13leading_zeros17h3d4a218ee1d90f6eE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN59_$LT$u32$u20$as$u20$compiler_builtins..int..traits..Int$GT$11rotate_left17h1115f774ce9d9d0bE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN59_$LT$u32$u20$as$u20$compiler_builtins..int..traits..Int$GT$13leading_zeros17h6aaf72df37750b87E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN59_$LT$u64$u20$as$u20$compiler_builtins..int..traits..Int$GT$11rotate_left17h2f64a84eb0cfbd47E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN59_$LT$u64$u20$as$u20$compiler_builtins..int..traits..Int$GT$13leading_zeros17hf4da95009ffec7fdE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN60_$LT$u128$u20$as$u20$compiler_builtins..int..traits..Int$GT$11rotate_left17h01a1b3591b465efdE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN60_$LT$u128$u20$as$u20$compiler_builtins..int..traits..Int$GT$13leading_zeros17h826bdcfb4d648c4cE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN61_$LT$isize$u20$as$u20$compiler_builtins..int..traits..Int$GT$11rotate_left17h9a3d5953e671482fE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN61_$LT$isize$u20$as$u20$compiler_builtins..int..traits..Int$GT$13leading_zeros17h850b03d47bab9fefE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN61_$LT$isize$u20$as$u20$compiler_builtins..int..traits..Int$GT$5ilog217hb687efc8b5dd77ebE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN61_$LT$usize$u20$as$u20$compiler_builtins..int..traits..Int$GT$11rotate_left17h59c7ac6f2615238cE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN61_$LT$usize$u20$as$u20$compiler_builtins..int..traits..Int$GT$13leading_zeros17h8c5b174b3219156fE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN61_$LT$usize$u20$as$u20$compiler_builtins..int..traits..Int$GT$5ilog217h973fc265ebc98b72E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN78_$LT$u8$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$11rotate_left17h7704e8d17e2e7624E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN78_$LT$u8$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$13leading_zeros17hb4ff14684367c625E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN79_$LT$u16$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$11rotate_left17h8aa85924e649c7c3E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN79_$LT$u16$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$13leading_zeros17he08f8988e6f0e76eE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN79_$LT$u32$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$11rotate_left17hddb309d143d83f66E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN79_$LT$u32$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$13leading_zeros17h399538a2271ebefeE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN79_$LT$u64$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$11rotate_left17h7e46279ce5fd4cbbE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN79_$LT$u64$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$13leading_zeros17h6c6ff17cb280bc08E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN80_$LT$u128$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$11rotate_left17h8e7408c10c9befbaE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN80_$LT$u128$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$13leading_zeros17h9e38ed835ed9abcaE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN81_$LT$isize$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$11checked_add17h7ab61916e1276626E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN81_$LT$isize$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$11checked_sub17h4662ba6789d22995E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN81_$LT$isize$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$11rotate_left17he9548172e0fa0a6fE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN81_$LT$isize$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$13leading_zeros17hd4127398ae10d924E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN81_$LT$isize$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$3abs17h694bc9ecea253edbE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN81_$LT$isize$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$5ilog217h0284336a96457fd0E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN81_$LT$usize$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$11checked_add17h86b874e3c1fc177bE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN81_$LT$usize$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$11rotate_left17hedf8a5e00a4eabceE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN81_$LT$usize$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$13leading_zeros17h06f1c071d37cb537E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN81_$LT$usize$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$3abs17hf221d017495e234dE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN81_$LT$usize$u20$as$u20$compiler_builtins..math..libm..support..int_traits..Int$GT$5ilog217h9edc3984e174ac96E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN84_$LT$compiler_builtins..int..big..u256$u20$as$u20$core..ops..bit..Shl$LT$u32$GT$$GT$3shl17he811c52996073932E (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN90_$LT$compiler_builtins..int..big..u256$u20$as$u20$compiler_builtins..int..traits..DInt$GT$2hi17hf5bff10ef16850ccE (EC Symbol)␍
          symbols.o : error LNK2001: unresolved external symbol #_ZN90_$LT$compiler_builtins..int..big..u256$u20$as$u20$compiler_builtins..int..traits..DInt$GT$2lo17h5f4e3db3104e3b3eE (EC Symbol)␍
          D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2-std\arm64ec-pc-windows-msvc\release\deps\std-23f1be5f370902f3.dll : fatal error LNK1120: 58 unresolved externals␍
          

[RUSTC-TIMING] std test:false 20.809
error: could not compile `std` (lib) due to 1 previous error
Build completed unsuccessfully in 1:47:55
```

</details>

--- a/compiler/rustc_codegen_ssa/src/back/linker.rs
+++ b/compiler/rustc_codegen_ssa/src/back/linker.rs
@@ -1782,7 +1782,10 @@ fn exported_symbols_for_non_proc_macro(tcx: TyCtxt<'_>, crate_type: CrateType) -
     let mut symbols = Vec::new();
     let export_threshold = symbol_export::crates_export_threshold(&[crate_type]);
     for_each_exported_symbols_include_dep(tcx, crate_type, |symbol, info, cnum| {
-        if info.level.is_below_threshold(export_threshold) {
+        // Do not export mangled symbols from cdylibs and don't attempt to export compiler-builtins
+        // from any cdylib. The latter doesn't work anyway as we use hidden visibility for
+        // compiler-builtins. Most linkers silently ignore it, but ld64 gives a warning.
+        if info.level.is_below_threshold(export_threshold) && !tcx.is_compiler_builtins(cnum) {
             symbols.push(symbol_export::exporting_symbol_name_for_instance_in_crate(
                 tcx, symbol, cnum,
             ));

--- a/compiler/rustc_codegen_ssa/src/back/linker.rs
+++ b/compiler/rustc_codegen_ssa/src/back/linker.rs
@@ -1824,7 +1824,9 @@ pub(crate) fn linked_symbols(
 
     let export_threshold = symbol_export::crates_export_threshold(&[crate_type]);
     for_each_exported_symbols_include_dep(tcx, crate_type, |symbol, info, cnum| {
-        if info.level.is_below_threshold(export_threshold) || info.used {
+        if info.level.is_below_threshold(export_threshold) && !tcx.is_compiler_builtins(cnum)
+            || info.used
+        {
             symbols.push((
                 symbol_export::linking_symbol_name_for_instance_in_crate(tcx, symbol, cnum),
                 info.kind,

